### PR TITLE
Robust error handling

### PR DESCRIPTION
**Pull request by dhruvan2006**
If you find a single `unwrap`, `expect`, `let _ = `, `Err("")`, `Box<dyn std::error::Error>`, `panic!()`, `break`, `if let Some(_)` you get 10 bitcoin.

### Familiarization:
1. Look first at the new enum Errors in main.rs
2. Look at `handle_request()` where I do case matching
3. Look commit by commit when reviewing, I've tried organizing it well.

### Key points:
- Only used `CliError` if both api error and cred error would occur. 
- Any error from `ApiError` and `CredentialError` that happens inside `handle_request()` will be handled there.
- Any other error in `CliError` like `IOError`, `JSONError`, `ConfigError` or some other in main.rs **will** cause the program to stop.

### Questions:
- When running `tuenroll start` offline should I exit immediately or let it loop every 5 seconds.
- How should we handle offline (currently we spam every 5 seconds)

### CODE REVIEW

**Comment by Rodvdc** [comment]
When running tuenroll start offline should I exit immediately or let it loop every 5 seconds.
- Probably exit immediately and tell them to try again later due to network error.
How should we handle offline (currently we spam every 5 seconds)
- Same type but perhaps a longer sleep time

**Comment by Rodvdc** [comment]
CliError::ApiError(ApiError::NetworkError(e))) can just be returned as ApiError::NetworkError(e)). Which might be easier to read.

File: src/controller.rs [hunk 1]
```diff
@@ -521,19 +584,27 @@ mod tests {
...
             };
-            let result = controller.run_auto_sign_up(|_| {}).await;
+            let result = controller.run_auto_sign_up().await;
 
             assert!(result.is_err());
-            assert_eq!(result.unwrap_err(), "Invalid credentials");
+            match result.unwrap_err() {
+                CliError::CredentialError(CredentialError::InvalidCredentials) => {}
+                _ => panic!("Expected InvalidCredentials error"),
+            }
```
Comment (Rodvdc) [inline src/controller.rs:607]: Shouldn't you be using assertEq!()?

File: src/creds.rs [hunk 2]
```diff
@@ -549,31 +571,34 @@ mod tests {
         let result = manager.retrieve_new_access_token(&mut credentials).await;
 
         assert!(result.is_err());
-        assert_eq!(result.unwrap_err().to_string(), "Invalid credentials");
+        match result.unwrap_err() {
+            CliError::CredentialError(CredentialError::InvalidCredentials) => {}
+            _ => panic!("Expected InvalidCredentials error"),
+        }
```
Comment (Rodvdc) [inline src/creds.rs:577]: Same again, shouldn't this be assertEq()

File: src/creds.rs [hunk 3]
```diff
@@ -549,31 +571,34 @@ mod tests {
...
+    //         password: Some("test_password".to_string()),
+    //         access_token: None,
+    //     };
+    //
+    //     let manager = CredentialManager::new(mock_api, "test_service".to_string());
+    //     let result = manager.retrieve_new_access_token(&mut credentials).await;
+    //
+    //     assert!(result.is_err());
+    //     assert_eq!(result.unwrap_err().to_string(), "Network request error");
+    // }
```
Comment (Rodvdc) [inline src/creds.rs:601]: Should this be added back?

Comment (Rodvdc) [reply src/creds.rs:601]: Same goes for line 920 - 946

File: src/controller.rs (hunk 1 above)
Comment (dhruvan2006) [reply src/controller.rs:601]: Have to make the enum derive PartialEq then

**Pull request merged**